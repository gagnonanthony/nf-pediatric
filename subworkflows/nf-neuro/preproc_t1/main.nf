// ** Importing modules from nf-neuro ** //
include { DENOISING_NLMEANS } from '../../../modules/nf-neuro/denoising/nlmeans/main'
include { PREPROC_N4 } from '../../../modules/nf-neuro/preproc/n4/main'
include { IMAGE_RESAMPLE } from '../../../modules/nf-neuro/image/resample/main'
include { BETCROP_ANTSBET } from '../../../modules/nf-neuro/betcrop/antsbet/main'
include { BETCROP_SYNTHBET} from '../../../modules/nf-neuro/betcrop/synthbet/main'
include { BETCROP_CROPVOLUME as BETCROP_CROPVOLUME_T1 } from '../../../modules/nf-neuro/betcrop/cropvolume/main'
include { BETCROP_CROPVOLUME as BETCROP_CROPVOLUME_MASK } from '../../../modules/nf-neuro/betcrop/cropvolume/main'

params.run_synthbet = false

workflow PREPROC_T1 {

    take:
        ch_image            // channel: [ val(meta), image ]
        ch_template         // channel: [ val(meta), template ]                              , optional
        ch_probability_map  // channel: [ val(meta), probability-map, mask, initial-affine ] , optional
        ch_mask_nlmeans     // channel: [ val(meta), mask ]                                  , optional
        ch_ref_n4           // channel: [ val(meta), ref, ref-mask ]                         , optional
        ch_ref_resample     // channel: [ val(meta), ref ]                                   , optional
        ch_weights          // channel: [ val(meta), weights ]                               , optional

    main:

        ch_versions = Channel.empty()

        // ** Denoising ** //
        // Result : [ meta, image, mask | [] ]
        //  Steps :
        //   - join [ meta, image, mask | null ]
        //   - map  [ meta, image, mask | [] ]
        ch_nlmeans = ch_image
            .join(ch_mask_nlmeans, remainder: true)
            .map{ it[0..1] + [it[2] ?: []] }

        DENOISING_NLMEANS ( ch_nlmeans )
        ch_versions = ch_versions.mix(DENOISING_NLMEANS.out.versions.first())

        // ** N4 correction ** //
        // Result : [ meta, image, reference | [], mask | [] ]
        //  Steps :
        //   - join [ meta, image ] + [ reference, mask ] | [ reference, null ] | [ null ]
        //   - map  [ meta, image, reference | [], mask | [] ]
        //   - join [ meta, image, reference | [], mask | [], nlmeans-mask | null ]
        //   - map  [ meta, image, reference | [], mask | [] ]
        ch_N4 = DENOISING_NLMEANS.out.image
            .join(ch_ref_n4, remainder: true)
            .map{ it[0..1] + [it[2] ?: [], it[3] ?: []] }
            .join(ch_mask_nlmeans, remainder: true)
            .map{ it[0..2] + [it[3] ?: it[4] ?: []] }

        PREPROC_N4 ( ch_N4 )
        ch_versions = ch_versions.mix(PREPROC_N4.out.versions.first())

        // ** Resampling ** //
        // Result : [ meta, image, reference | [] ]
        //  Steps :
        //   - join [ meta, image, reference | null ]
        //   - map  [ meta, image, reference | [] ]
        ch_resampling = PREPROC_N4.out.image
            .join(ch_ref_resample, remainder: true)
            .map{ it[0..1] + [it[2] ?: []] }

        IMAGE_RESAMPLE ( ch_resampling )
        ch_versions = ch_versions.mix(IMAGE_RESAMPLE.out.versions.first())

        // ** Brain extraction ** //
        if ( params.run_synthbet ) {
            // ** SYNTHBET ** //
            // Result : [ meta, image, weights | [] ]
            //  Steps :
            //   - join [ meta, image, weights | null ]
            //   - map  [ meta, image, weights | [] ]
            ch_bet = IMAGE_RESAMPLE.out.image
                .join(ch_weights, remainder: true)
                .map{ it[0..1] + [it[2] ?: []] }

            BETCROP_SYNTHBET ( ch_bet )
            ch_versions = ch_versions.mix(BETCROP_SYNTHBET.out.versions.first())

            // ** Setting BET output ** //
            image_bet = BETCROP_SYNTHBET.out.bet_image
            mask_bet = BETCROP_SYNTHBET.out.brain_mask
        }

        else {
            // ** ANTSBET ** //
            // The template and probability maps are mandatory if running antsBET. Since the
            // error message from nextflow when they are absent is either non-informative or
            // missing, we use ifEmpty to provide a more informative one.
            // Remove this error message for now, as it is buggy.
            ch_bet = IMAGE_RESAMPLE.out.image
                .join(ch_template)
                .join(ch_probability_map)
                .map{ it + [[], []] }

            BETCROP_ANTSBET ( ch_bet )
            ch_versions = ch_versions.mix(BETCROP_ANTSBET.out.versions.first())

            // ** Setting BET output ** //
            image_bet = BETCROP_ANTSBET.out.t1
            mask_bet = BETCROP_ANTSBET.out.mask
        }

        // ** Crop image ** //
        ch_crop = image_bet
            .map{ it + [[]] }

        BETCROP_CROPVOLUME_T1 ( ch_crop )
        ch_versions = ch_versions.mix(BETCROP_CROPVOLUME_T1.out.versions.first())

        // ** Crop mask ** //
        ch_crop_mask = mask_bet
            .join(BETCROP_CROPVOLUME_T1.out.bounding_box)

        BETCROP_CROPVOLUME_MASK ( ch_crop_mask )
        ch_versions = ch_versions.mix(BETCROP_CROPVOLUME_MASK.out.versions.first())

    emit:
        t1_final        = BETCROP_CROPVOLUME_T1.out.image           // channel: [ val(meta), t1-preprocessed ]
        mask_final      = BETCROP_CROPVOLUME_MASK.out.image         // channel: [ val(meta), t1-mask ]
        image_nlmeans   = DENOISING_NLMEANS.out.image               // channel: [ val(meta), t1-after-denoise ]
        image_N4        = PREPROC_N4.out.image                      // channel: [ val(meta), t1-after-unbias ]
        image_resample  = IMAGE_RESAMPLE.out.image                  // channel: [ val(meta), t1-after-resample ]
        image_bet       = image_bet                                 // channel: [ val(meta), t1-after-bet ]
        mask_bet        = mask_bet                                  // channel: [ val(meta), intermediary-mask ]
        crop_box        = BETCROP_CROPVOLUME_T1.out.bounding_box    // channel: [ val(meta), bounding-box ]
        versions        = ch_versions                               // channel: [ versions.yml ]
}
