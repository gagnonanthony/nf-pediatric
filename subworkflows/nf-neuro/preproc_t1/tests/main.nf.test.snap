{
    "preproc_t1_synthbet": {
        "content": [
            "test__denoised.nii.gz:md5:header,a7ee0e819007aba98d14f7a145b550e6,data,2e21433e2bcd4de2a6b6167c6233cd40",
            "test__image_n4.nii.gz:md5:header,e7cfbd06624321d70cbd667a77315ba3,data,a81e98f32ed963c098ccb07486101898",
            "test_resampled.nii.gz:md5:header,7628a07204938d640c3530fa3d76d2b7,data,a81e98f32ed963c098ccb07486101898",
            "test__bet_image.nii.gz:md5:header,38a09a5addf7c1f13dae1121c562f3b5,data,74f76699e09839809038875af1d9ae6c",
            "test__brain_mask.nii.gz:md5:header,8fc2b4ae979d881623dfd34d377d437d,data,2d61506dc4ab2f8093b731474ae7e45c",
            [
                [
                    {
                        "id": "test",
                        "single_end": false
                    },
                    "test_t1_cropped_bbox.pkl:md5,522d2c44d3ad1058ea77457a263e39c8"
                ]
            ],
            "test_cropped.nii.gz:md5:header,efab188f3700b5b29d4b4ef99cec1295,data,e551be653d402018fd73a4f708b1641e",
            "test_t1_cropped.nii.gz:md5:header,efab188f3700b5b29d4b4ef99cec1295,data,8a93b9d76ead7dae4af4792b9cf70479",
            [
                "versions.yml:md5,318cabe934be45528a25f52083d9c90d",
                "versions.yml:md5,37229cd5cee6f6788fbc9083d6618c61",
                "versions.yml:md5,657fbb224f260392e573b8511c4b798d",
                "versions.yml:md5,b979132991d8f72a3585465533bd5730",
                "versions.yml:md5,bdd934b4b8456060c36d6d97e4f30740",
                "versions.yml:md5,ea32c30f5320f720b2f5dc32ac2535ea"
            ]
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "24.10.0"
        },
        "timestamp": "2024-11-25T18:23:03.990173398"
    },
    "preproc_t1_antsbet": {
        "content": [
            "test__denoised.nii.gz:md5:header,a7ee0e819007aba98d14f7a145b550e6,data,2e21433e2bcd4de2a6b6167c6233cd40",
            "test__image_n4.nii.gz:md5:header,e7cfbd06624321d70cbd667a77315ba3,data,a81e98f32ed963c098ccb07486101898",
            "test_resampled.nii.gz:md5:header,7628a07204938d640c3530fa3d76d2b7,data,a81e98f32ed963c098ccb07486101898",
            "test__t1_bet.nii.gz",
            "test__t1_bet_mask.nii.gz",
            "test_t1_cropped_bbox.pkl",
            "test_cropped.nii.gz",
            "test_t1_cropped.nii.gz",
            [
                "versions.yml:md5,19aaa24cf046e49a4ad4ce5d507ea8aa",
                "versions.yml:md5,37229cd5cee6f6788fbc9083d6618c61",
                "versions.yml:md5,657fbb224f260392e573b8511c4b798d",
                "versions.yml:md5,b979132991d8f72a3585465533bd5730",
                "versions.yml:md5,bdd934b4b8456060c36d6d97e4f30740",
                "versions.yml:md5,ea32c30f5320f720b2f5dc32ac2535ea"
            ]
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.04.3"
        },
        "timestamp": "2024-12-04T10:38:12.185789"
    }
}